vllm/distributed/device_communicators/custom_all_reduce.py

```diff
@@ -149,6 +149,13 @@ def __init__(self,
         dist.all_gather(gather_list, tensor, group=self.group)
         physical_device_ids = [t.item() for t in gather_list]
 
+        # 3 nodes will be like [0,1,2,3,0,1,0,1]
+        if sorted(physical_device_ids)[-1] + 1 != world_size:
+            logger.warning(
+                "Custom allreduce is disabled because this feature is "
+                "not intended for multi node use case.")
+            return
+
```

Comment:
This is still not safe. It is possible, that one node uses gpu `0, 1` , while the other node uses `2, 3`, and the world size is `4`. In this case, we cannot use custom allreduce. However, this code will not capture it.